core/src/nucleus/mod.rs

```diff
@@ -183,72 +186,85 @@ fn reduce_ia(
-                // initialization result on its own, because then it will miss the genesis
-                // sometimes where a genesis does happen.
-                if results.is_empty() {
-                    results.push(CallbackResult::Pass);
-                }
+    // check pre-condition
+    if state.status() != NucleusStatus::New {
+        // Send bad start state ReturnInitializationResult Action
+        return_initialization_result(
+            Some("Nucleus already initialized or initializing".to_string()),
+            &action_channel,
+        );
+        return;
+    }
 
-                // map the genesis results to initialization result responses
-                for result in results {
-                    match result {
-                        CallbackResult::Fail(s) => return_initialization_result(
-                            Some(s.to_string()),
-                            &genesis_action_channel,
-                        ),
-                        _ => return_initialization_result(None, &genesis_action_channel),
-                    }
-                }
-            });
-        }
-        _ => {
-            // Send bad start state ReturnInitializationResult Action
-            return_initialization_result(
-                Some("Nucleus already initialized or initializing".to_string()),
-                &action_channel,
+    let ia_action = action_wrapper.action();
+    let dna = unwrap_to!(ia_action => Action::InitApplication);
+
+    // Update status
+    state.status = NucleusStatus::Initializing;
+
+    // Set DNA
+    state.dna = Some(dna.clone());
+
+    // Create & launch thread
+    let genesis_action_channel = action_channel.clone();
+    let genesis_observer_channel = observer_channel.clone();
+    let dna_clone = dna.clone();
+    // let action_wrapper_clone = action_wrapper.clone();
+
+    thread::spawn(move || {
+        // Send Commit Action for Genesis Entry
+        {
+            // Create Commit Action for Genesis Entry
+            let genesis_entry = dna_clone.to_entry();
+            let commit_genesis_action = ActionWrapper::new(Action::Commit(genesis_entry));
+
+            // Send Action and wait for it
+            ::instance::dispatch_action_and_wait(
+                &genesis_action_channel,
+                &genesis_observer_channel,
+                commit_genesis_action,
             );
         }
-    }
+
+        // map genesis across every zome
+        let mut results: Vec<_> = dna_clone
+            .zomes
+            .iter()
+            .map(|zome| {
+                genesis(
+                    &genesis_action_channel,
+                    &genesis_observer_channel,
+                    &zome.name(),
+                    &CallbackParams::Genesis,
+                )
+            })
+            .collect();
+
+        // pad out a single pass if there are no zome results
+        // @TODO is this really OK?
```

Comment:
oops i forgot to put a followup for this it seems, could you throw one in for me?

Reply:
I think there was already a card for that: https://waffle.io/holochain/org/cards/5b3644a6eaa711001ce607ce
Does that work for you? You can edit the card if you want.